Assigning to optional give c error: expected struct or union but not 'struct main__IGameObject *'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

<details>
<summary>V version: V 0.4.10 b76b1ee, press to see full `v doctor` output</summary>

|V full version      |V 0.4.10 5c65e58.b76b1ee
|:-------------------|:-------------------
|OS                  |windows, Microsoft Windows 11 Pro 26100 64-bit
|Processor           |24 cpus, 64bit, little endian, AMD Ryzen 9 3900X 12-Core Processor  
|Memory              |11.33GB/31.9GB
|                    |
|V executable        |C:\v\v.exe
|V last modified time|2025-04-22 07:58:32
|                    |
|V home dir          |OK, value: C:\v
|VMODULES            |OK, value: C:\Users\info\.vmodules
|VTMP                |OK, value: C:\Users\info\AppData\Local\Temp\v_0
|Current working dir |OK, value: D:\vbug2
|                    |
|Git version         |git version 2.42.0.windows.2
|V git status        |weekly.2025.14-82-gb76b1ee7-dirty
|.git/config present |true
|                    |
|cc version          |N/A
|gcc version         |N/A
|clang version       |clang version 17.0.1
|msvc version        |N/A
|tcc version         |tcc version 0.9.27 (x86_64 Windows)
|tcc git status      |thirdparty-windows-amd64 b425ac82
|emcc version        |N/A
|glibc version       |ldd (cygwin) 3.4.7

</details>

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg  main.v && D:\vbug2\main.exe`
```v
module main

@[heap]
interface IGameObject {
mut:
	name       string
	parent     ?&IGameObject
	next       ?&IGameObject
	child      ?&IGameObject
	last_child ?&IGameObject
	add_child(mut o IGameObject)
}

@[heap]
struct GameObject implements IGameObject {
mut:
	name       string
	parent     ?&IGameObject
	next       ?&IGameObject
	child      ?&IGameObject
	last_child ?&IGameObject
}

fn (mut gameobject GameObject) add_child(mut o IGameObject) {
	o.parent = gameobject
	if gameobject.last_child != none {
		gameobject.last_child.next = o
	} else {
		gameobject.child = o
	}
	gameobject.last_child = o
}

fn main() {
	mut v1 := &GameObject{
		name: 'v1'
	}
	mut v2 := &GameObject{
		name: 'v2'
	}
	v1.add_child(mut v2)
}

```

**What did you see?**
```
================== C compilation error (from tcc): ==============
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JSE8KMQVRCSRQAW4QH7519FF.tmp.c:3852: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JSE8KMQVRCSRQAW4QH7519FF.tmp.c:5852: error: expected struct or union but not 'struct main__IGameObject *'
...
cc: C:/Users/info/AppData/Local/Temp/v_0/main.01JSE8KMQVRCSRQAW4QH7519FF.tmp.c:5852: error: expected struct or union but not 'struct main__IGameObject *'
(note: the original output was 3 lines long; it was truncated to its first 2 lines + the last line)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

**What did you expect to see?**

no c error



> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -153,6 +153,7 @@ mut:
 	inside_cast_in_heap       int // inside cast to interface type in heap (resolve recursive calls)
 	inside_cast               bool
 	inside_selector           bool
+	inside_selector_deref     bool // indicates if the inside selector was already dereferenced
 	inside_memset             bool
 	inside_const              bool
 	inside_array_item         bool
@@ -4173,9 +4174,9 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 							deref := if g.inside_selector {
 								if is_iface_or_sumtype || (field.orig_type.is_ptr() && g.left_is_opt
 									&& is_option_unwrap) {
-									'*'.repeat(field.smartcasts.last().nr_muls())
+									'*'.repeat(typ.nr_muls())
 								} else {
-									'*'.repeat(field.smartcasts.last().nr_muls() + 1)
+									'*'.repeat(typ.nr_muls() + 1)
 								}
 							} else if sym.kind == .interface && !typ.is_ptr()
 								&& field.orig_type.has_flag(.option) {
@@ -4183,7 +4184,14 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 							} else {
 								'*'
 							}
-							g.write('(${deref}(${g.styp(typ)}*)')
+							suffix := if g.inside_selector && field.orig_type.has_flag(.option)
+								&& field.orig_type.is_ptr() && g.is_assign_lhs {
+								''
+							} else {
+								'*'
+							}
+							g.inside_selector_deref = suffix == '' && deref.len > 0
+							g.write('(${deref}(${g.styp(typ)}${suffix})')
 						}
 						if i == 0 || !nested_unwrap {
 							g.write('(')
@@ -4289,6 +4297,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		g.write('((${g.base_type(field_typ)})')
 	}
 	old_inside_selector := g.inside_selector
+	old_inside_selector_deref := g.inside_selector_deref
 	g.inside_selector = node.expr is ast.SelectorExpr && node.expr.expr is ast.Ident
 	n_ptr := node.expr_type.nr_muls() - 1
 	if n_ptr > 0 {
@@ -4298,7 +4307,9 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 	} else {
 		g.expr(node.expr)
 	}
+	mut opt_ptr_already_deref := g.inside_selector_deref
 	g.inside_selector = old_inside_selector
+	g.inside_selector_deref = old_inside_selector_deref
 	if field_is_opt {
 		g.write(')')
 	}
@@ -4350,7 +4361,7 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 		|| (((!is_dereferenced && unwrapped_expr_type.is_ptr()) || sym.kind == .chan
 		|| alias_to_ptr) && node.from_embed_types.len == 0)
 		|| (node.expr.is_as_cast() && g.inside_smartcast)
-	if !has_embed && left_is_ptr {
+	if !has_embed && left_is_ptr && !(opt_ptr_already_deref && !g.inside_selector) {
 		g.write('->')
 	} else {
 		g.write('.')

--- a/vlib/v/tests/options/option_selector_unwrapping_test.v
+++ b/vlib/v/tests/options/option_selector_unwrapping_test.v
@@ -1,0 +1,44 @@
+module main
+
+@[heap]
+interface IGameObject {
+mut:
+	name       string
+	parent     ?&IGameObject
+	next       ?&IGameObject
+	child      ?&IGameObject
+	last_child ?&IGameObject
+	add_child(mut o IGameObject)
+}
+
+@[heap]
+struct GameObject implements IGameObject {
+mut:
+	name       string
+	parent     ?&IGameObject
+	next       ?&IGameObject
+	child      ?&IGameObject
+	last_child ?&IGameObject
+}
+
+fn (mut gameobject GameObject) add_child(mut o IGameObject) {
+	o.parent = gameobject
+	if gameobject.last_child != none {
+		gameobject.last_child.next = o
+	} else {
+		gameobject.child = o
+	}
+	gameobject.last_child = o
+}
+
+fn test_main() {
+	mut v1 := &GameObject{
+		name: 'v1'
+	}
+	mut v2 := &GameObject{
+		name: 'v2'
+	}
+	v1.add_child(mut v2)
+	assert v1.child? == IGameObject(v2)
+	assert v1.last_child? == IGameObject(v2)
+}

